feat(curriculum): add string method workshop (#66936)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-employee-profile-validator/69e20a0451c7a3ef189514d3.md

---
id: 69e20a0451c7a3ef189514d3
title: Step 4
challengeType: 20
dashedName: step-4
---

# --description--

> Now that you know the position of `@` in the email, you can extract the username, which is placed before the @ sign.

Reviewer: ```suggestion
Now that you know the position of `@` in the email, you can extract the username, which is placed before the `@` sign.
```

In the previous workshop, you extracted a portion of a string using slicing. As a reminder, the syntax is:

> 

Reviewer: ````suggestion

```py
string[start:end]
```

Now, use slicing with `at_position` to extract the part of `email` that comes before the `@` sign, and assign the result to a variable named `username`.
Then print the result of concatenating the string `Username: ` and  the `username` variable.

# --hints--

````

You should create a variable named `username`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('username')`)) })
```

You should use slicing with `at_position` on `email` to assign `username` a value.

```js
({ test: () => assert(runPython(`
node = _Node(_code).find_variable('username')
not node.is_empty() and node.is_equivalent("username = email[:at_position]")
`)) })
```

> You should print `username` with the prefix `'Username: '`.

Reviewer: ```suggestion
You should print the result of concatenating the string `Username: ` and the `username` variable.
```

---
id: 69e20a0451c7a3ef189514d3
title: Step 4
challengeType: 20
dashedName: step-4
---

# --description--

Now that you know the position of `@` in the email, you can extract the username, which is placed before the @ sign.

In the previous workshop, you extracted a portion of a string using slicing. As a reminder, the syntax is:

> # --hints--

Reviewer 2: The description is cut off after "the syntax is:", so the promised slicing example is missing and the step never states the task. The three hints below grade `username`, the slice `email[:at_position]`, and a print with the prefix `'Username: '`, none of which the prose asks for.

`````suggestion
In the previous workshop, you extracted a portion of a string using slicing. As a reminder, the syntax is:

```py
text = 'hello world'
print(text[:5])  # hello
```

Now, use slicing with `at_position` to extract the part of `email` that comes before the `@` sign, and assign the result to a variable named `username`.

Then, print `username` with the prefix `'Username: '`.

# --hints--
`````

You should create a variable named `username`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('username')`)) })
```

> You should use slicing with `at_position` on `email` to assign a value to `username`.

Reviewer 2: The majority form across the workshop is "to assign `X` a value" (steps 11, 12, 13, 14, 17, 18, 20).

```suggestion
You should use slicing with `at_position` on `email` to assign `username` a value.
```

---
id: 69e20a0451c7a3ef189514d3
title: Step 4
challengeType: 20
dashedName: step-4
---

# --description--

Now that you know the position of `@` in the email, you can extract the username, which is placed before the `@` sign.

> In the previous workshop, you extracted a portion of a string using slicing. As a reminder, the syntax is:

Reviewer 2: `@` terminology on L10, see the step 3 note.

L12 says "In the previous workshop" where step 2 says "Remember from a previous workshop" about the same earlier material. Using "a previous workshop" in both keeps them consistent and stays true if the ordering ever changes.

```suggestion
Now that you know the position of `@` in the email, you can extract the username, which is placed before the `@` character.

In a previous workshop, you extracted a portion of a string using slicing. As a reminder, the syntax is:
```

```python
string[start:end]
```

If you leave out `start`, it defaults to `0`. For example:

```python
text = 'hello world'
first_word = text[:5]
print(first_word)  # Output: hello
```

> Use slicing on `email` to extract everything before `'@'`. Store the result in a variable named `username` and print it with the prefix `'Username: '`.

Reviewer 3: ````suggestion
In the previous workshop, you extracted a portion of a string using slicing. As a reminder, the syntax is:

```py
string[start:end]
```

Use string slicing on the `email` variable to extract the username. Use the `at_position` variable as the `end` index and remember that you can leave out the `start` index because it defaults to `0`.

Store the result of the slicing in a variable named `username`. Then print the result of concatenating the string `'Username: '` and the `username` variable.
````

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('username')`)) })
```

You should use slicing with `at_position` on `email` to assign `username` a value.

```js
({ test: () => assert(runPython(`
node = _Node(_code).find_variable('username')

> not node.is_empty() and node.is_equivalent("username = email[:at_position]")

Reviewer 2: This hint accepts only `email[:at_position]`, so `email[0:at_position]` is rejected even though the description just says "extract everything before `at_position`". Step 12 grades the identical operation and accepts both spellings, so the two steps disagree about the same answer:

```python
Node("u = email[0:at]").find_variable('u').is_equivalent("u = email[:at]")  # False
```

The suggestion mirrors step 12's hint. `not node.is_empty() and` is dropped for the reason in the step 5 comment, and the prose line above would read better as step 12's wording, which the phrasing comment on step 3 covers.

```suggestion
node = _Node(_code).find_variable('username')
node.is_equivalent("username = email[:at_position]") or node.is_equivalent("username = email[0:at_position]")
```
